src/Microsoft.Identity.Web/Resource/TokenAuthorizationRules.cs

```diff
@@ -0,0 +1,150 @@
+        /// <param name="claimsPrincipal">Claims in the token.</param>
+        public static void ValidateApiAuthorization(
+            ITokenAuthorizationOptions tokenAuthorizationOptions,
+            ClaimsPrincipal claimsPrincipal)
+        {
+            if (tokenAuthorizationOptions is null)
+            {
+                throw new ArgumentNullException(nameof(tokenAuthorizationOptions));
+            }
+
+            if (claimsPrincipal is null)
+            {
+                throw new ArgumentNullException(nameof(claimsPrincipal));
+            }
+
+            // Does the web API disallow app only tokens?
+            if (!tokenAuthorizationOptions.ApiAllowsAppOnlyTokens)
+            {
+                VerifyClaimDoesNotExistOrDoesNotEquals(claimsPrincipal, ClaimConstants.Idtyp, "app", IDWebErrorMessage.ApiAllowsAppOnlyTokensCantBeComputed, IDWebErrorMessage.ApiDoesNotAllowsAppOnlyTokens);
+            }
+
+            // Does the web API disallow guest accounts?
+            if (!tokenAuthorizationOptions.ApiAllowsGuestAccounts)
+            {
+                VerifyClaimEquals(claimsPrincipal, ClaimConstants.Acct, "0", IDWebErrorMessage.ApiAllowsGuestAccountsCantBeComputed, IDWebErrorMessage.ApiDoesNotAllowsGuestAccounts);
+            }
+
+            // Does the web API demand specific tenants?
+            if (tokenAuthorizationOptions.AllowedTenantIds != null)
+            {
+                VerifyClaimIsInAcceptedValues(
+                    claimsPrincipal,
+                    new[] { ClaimConstants.Tid, ClaimConstants.TenantId },
+                    tokenAuthorizationOptions.AllowedTenantIds,
+                    IDWebErrorMessage.ApiAllowedTenantsCantBeComputed,
+                    IDWebErrorMessage.ApiTenantNotAllowed);
+            }
+
+            // Does the web API demand specific clients (oid)
+            if (tokenAuthorizationOptions.AllowedClientApplications != null)
+            {
+                VerifyClaimIsInAcceptedValues(
+                    claimsPrincipal,
+                    new[] { ClaimConstants.Oid, ClaimConstants.ObjectId },
+                    tokenAuthorizationOptions.AllowedClientApplications,
+                    IDWebErrorMessage.ApiAllowedClientApplicationsCantBeComputed,
+                    IDWebErrorMessage.ApiClientApplicationNotAllowed);
+            }
+        }
+
+        /// <summary>
+        /// Verify the value of a claim.
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims.</param>
+        /// <param name="claimName">Name (type) of the claim to verify.</param>
+        /// <param name="expectedValue">Expected value.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimEquals(
```

Comment:
This throws an exception fi the claim _does_ equal the expected value. It should throw if it does _not_, right?

Reply:
Yes. (this is a draft PR for a proof of concept for the moment). Good catch, @psignoret

Reply:
Thanks @psignoret. Fixed